### DISCUSSION TITLE
Getting data out of loadContext

### DISCUSSION THREAD

**Discussion by rossipedia** [remix-run/remix#4603] [Q&A]
I'm using the Node/Express adaper, and let's say I have a utility that I want to expose to all my loaders:

**`server.ts`**
```ts
declare module '@remix-run/server-runtime' {
  interface AppLoadContext {
    doTheThing(): void;
  }
}

app.all(
  '*',
  createRequestHandler({
    build,
    mode: config.buildConfig,
    getLoadContext() {
      return {
        doTheThing() {
          ...
        }
      };
    },
  }),
);
```

**`app/routes/some-url.tsx`**:

```ts
export function loader({ request, context }) {
  // do the thing
  context.doTheThing();
  // ...
  return json(...);
}
```

Now, let's say that `context.doTheThing` collects information about how it's used. Maybe it's profiling information, maybe it's query data that's meant to be dehydrated.

My question is:

**_How do I get data out of the loader context into my `entry.server.tsx`?_**

The server entry default export looks like this:

```ts
export default function handleRequest(
  request: Request,
  responseStatusCode: number,
  responseHeaders: Headers,
  remixContext: EntryContext,
) {
 ...
}
```

And from following the source in `@remix-run/express` and `@remix-run/server-runtime`, it doesn't look like there's any direct path there. 

Maybe there's something I'm not seeing?

**Comment by kiliman** [comment]
In *entry.server*, the new APIs `handleDataRequest` and `handleResourceRequest` include the `loadContext` as a parameter. Unfortunately the default export for handling document request does not. The only way to do that currently is to patch *remix-server-runtime/server.ts* to include that parameter. I'm pretty sure the new APIs included it because they simply copied the signature of the internal function, so it kind of came along for the ride.

https://github.com/remix-run/remix/blob/8f0f0f6e1fbad4edaeee5a69bc2177b2e260a32c/packages/remix-server-runtime/server.ts#L610-L619

**Reply by rossipedia** [reply]
Well that's kinda ironic, seeing as I need to get that data from `loadContext` to the client. I saw there was a PR to add that param to the default entry export. Hopefully that can make it in and I can stop manually patching the server-runtime package.

**Reply by machour** [reply]
Link to the mentioned PR for additional context: https://github.com/remix-run/remix/pull/2045

**Reply by kiliman** [reply]
Right now, the only way to get that context info to the client is to return it in the loader.

**Reply by sergiodxa** [reply]
The problem is what happens if you use the AppLoadContext to inject code and not just data, you can't return data from a loader response, only JSON-compatible values.

**Reply by kentcdodds** [reply]
> Right now, the only way to get that context info to the client is to return it in the loader.

And this becomes problematic when the data you need is a nonce which should not go to the client 🙈 https://github.com/remix-run/remix/issues/5162

**Comment by rossipedia** [comment]
Per @jacob-ebey's suggest, I'm posting this use-case here that I gave on Discord: 

> It was originally an instrumentation goal. Let's say I have a utility that I want to expose to loaders on the context, but it also gathers information on how it's used, and then I want to output that data to the client in some form after all the loaders have executed (I realize entry.server.tsx is only one side of the solution here).
> 
> It's kinda the same thing as the middleware topic, but on the other end. A `afterLoaders` hook or similar. I don't know how it would work for loaders invoked by the client as opposed to the server request. But having access to the loaderContext in the server entry would be kinda cool to expose server request profiling information to the client in a rather straightforward way

And @sergiodxa followed up with:

> For loader invoked by the client you have `handleDataRequest` export in entry.server which does receive the context

The fact that `handleDataRequest` receives the loader context, but `handleRequest` in `entry.server.tsx` does not feels like an oversight. It means there are things you can do with one (a JSON data request from the client) that you can't do with the other (SSR full document request), even though both happen after all the loaders have been executed.

**Comment by kentcdodds** [comment]
I need this so I can access a `nonce` that https://npm.im/helmet is generating in my express server from within my entry. I need it in my entry so I can pass it along to `renderToPipeableStream` and my server-only `NonceProvider`. I've implemented a patch-package on Remix so I can do this and avoid the security issues: https://github.com/kentcdodds/kentcdodds.com/commit/f698a2dfd7f1dda321ec340d4ac4f645bfdf128b

I would *really* like to get rid of that patch...

More context: https://github.com/remix-run/remix/issues/5162

**Comment by ikibalnyi** [comment]
Need this functionality as well. Our use case is:
We detect user language in the `express` server part of the app during request. We need to access this language in the whole remix part of the app: the Root component, CatchBoundary to initialize the i18n provider. So since entry.X files is the only place that allows providing things to the entire app, we ended up passing it to `loadContext` and patching remix as @kentcdodds proposed.

**Comment by ikibalnyi** [comment]
@kentcdodds, I found a solution until it's fixed by the remix team. It's based on Node's `AsyncLocalStorage` available from `node@>=14` https://nodejs.org/api/async_hooks.html. Here I put it together in a gist:
https://gist.github.com/ikibalnyi/d23a3710f810ff0f434bf58c9039d7b3

**Reply by gustavopch** [reply]
Didn't work for me. I get the error `Error: Invariant: attempting to entryContext.getStore() outside of entryContext.run()`.

**Reply by gustavopch** [reply]
It didn't work for me because the `AsyncLocalStorage` instance used in my `server.ts` was different from the instance used in the loaders/actions when in development. One instance was constructed in my `app` folder and the other in `build/index.js`. I was able to get it to work by ensuring a single instance is used in the process:
```ts
export const asyncLocalStorage =
  (globalThis.asyncLocalStorage ??= new AsyncLocalStorage())
```

**Comment by sergiodxa** [comment]
This now implemented since v1.16.0 if I’m not wrong
